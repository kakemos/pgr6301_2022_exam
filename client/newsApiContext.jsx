import React from "react";
import { fetchJSON } from "./lib/fetchJSON";
import { postJSON } from "./lib/postJSON";
import { deleteJSON } from "./lib/deleteJSON";

export const NewsApiContext = React.createContext({
  async listFilteredArticles(query) {
    return await fetchJSON("/api/news?" + new URLSearchParams(query));
  },
  async addNewArticle(article) {
    return await postJSON("/api/news/new", article);
  },
  async fetchLogin() {
    return await fetchJSON("/api/login");
  },
  async registerLogin(provider, login) {
    return await postJSON(`/api/login/${provider}`, login);
  },
  async endSession() {
    const res = await fetch("/api/login", { method: "DELETE" });
    if (!res.ok) {
      throw new Error(`Failed to post ${res.status}: ${res.statusText}`);
    }
  },
  async showArticle(title) {
    return await fetchJSON("/api/news?" + new URLSearchParams(title));
  },
  async deleteArticle(title) {
    return await deleteJSON("/api/news/article?" + new URLSearchParams(title));
  },
});
