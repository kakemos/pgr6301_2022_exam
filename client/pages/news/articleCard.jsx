import React from "react";

export function ArticleCard({
  news: { title, intro, author, category, article },
}) {
  return (
    <>
      <div className={"article-card"}>
        <h1 className={"article-title"}>{title}</h1>
        <hr />
        <div className={"article-under-header"}>
          <h4>Written by {author}</h4>
          <h4>category: {category}</h4>
        </div>
        <p>{intro}</p>
        <p>{article}</p>
      </div>
    </>
  );
}
