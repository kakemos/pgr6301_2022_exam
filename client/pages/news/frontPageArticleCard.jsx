import { Link } from "react-router-dom";
import React from "react";

export function FrontPageArticleCard({ news: { title, intro, author } }) {
  console.log("front page article card title " + title);

  return (
    <>
      <div className={"front-page-article-card"}>
        <hr />
        <h2 className={"front-page-title"}>
          <Link to={"/article/" + title}>{title}</Link>
        </h2>
        <h3 className={"front-page-author"}>Written by {author}</h3>
        <p className={"front-page-intro"}>{intro}</p>
      </div>
    </>
  );
}
