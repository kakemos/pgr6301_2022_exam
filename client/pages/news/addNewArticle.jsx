import React, { useContext, useEffect, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useNavigate } from "react-router-dom";
import { FormInput, FormInputTextArea } from "../../lib/formInput";
import { useLoading } from "../../lib/useLoading";

export function AddNewArticle({ user }) {
  const { addNewArticle, listFilteredArticles } = useContext(NewsApiContext);
  const [title, setTitle] = useState("");
  const [intro, setIntro] = useState("");
  const [article, setArticle] = useState("");
  const [category, setCategory] = useState("general");
  const [author, setAuthor] = useState(user?.microsoft?.name);
  const [ws, setWs] = useState("");
  const navigate = useNavigate();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ title: title }),
    []
  );

  useEffect(() => {
    const ws = new WebSocket(window.location.origin.replace(/^http/, "ws"));
    setWs(ws);
  }, []);

  async function handleSubmit(event) {
    event.preventDefault();
    if (title === "" || intro === "" || article === "") {
      return navigate("/error-empty-fields");
    } else if (title === data.title) {
      return navigate("error-duplicate-title");
    } else {
      await addNewArticle({ title, intro, article, category, author });
      ws.send(JSON.stringify({ title, intro }));
      navigate(`/article/${title}`);
    }
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  if (!user.microsoft) {
    return <Navigate to="/login" replace />;
  }
  return (
    <>
      <h1>Write new article.</h1>
      <form onSubmit={handleSubmit}>
        <h4>Author: {author}</h4>
        <FormInput label={"Title"} value={title} onChangeValue={setTitle} />
        <FormInput label={"Intro"} value={intro} onChangeValue={setIntro} />
        <div className={"form-input"}>
          <div>
            <label htmlFor="dog-names">
              <strong>Category</strong>
            </label>
          </div>
          <select
            className={"new-article-dropdown"}
            name="categories"
            id="category-dropdown"
            value={category}
            onChange={(e) => setCategory(e.target.value)}
          >
            <option value="general">General</option>
            <option value="fashion">Fashion</option>
            <option value="art">Art</option>
            <option value="travel">Travel</option>
            <option value="celebrity">Celebrity</option>
          </select>
        </div>
        <FormInputTextArea
          label={"Article"}
          value={article}
          onChangeValue={setArticle}
        />
        <div>
          <button className={"submit-article-button"}>Submit article</button>
        </div>
      </form>
    </>
  );
}
