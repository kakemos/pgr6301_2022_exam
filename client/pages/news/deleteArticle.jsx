import React, { useContext, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useNavigate } from "react-router-dom";
import { useLoading } from "../../lib/useLoading";

export function DeleteArticle({ user }) {
  const { deleteArticle, listFilteredArticles } = useContext(NewsApiContext);
  const { title, setTitle } = useState("");
  const navigate = useNavigate();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ author: user?.microsoft?.name }),
    []
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  async function handleSubmit(event) {
    event.preventDefault();
    await deleteArticle({ title });
    navigate("/");
  }

  if (!user.microsoft) {
    return <Navigate to="/login" replace />;
  }

  return (
    <>
      <h2>Overview of all your articles.</h2>
      <h3>Which one do you wanna delete?</h3>
      <form onSubmit={handleSubmit}>
        {data.map((article) => (
          <>
            <input
              type={"radio"}
              name={"delete-articles"}
              value={article.title}
              onChange={() => setTitle(article.title)}
            />
            <label htmlFor={article.title}>{article.title}</label>
            <p>{article.intro}</p>
          </>
        ))}
        <button>Delete</button>
      </form>
    </>
  );
}
