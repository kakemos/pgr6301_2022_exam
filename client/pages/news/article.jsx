import React, { useContext } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { Navigate, useParams } from "react-router-dom";
import { useLoading } from "../../lib/useLoading";
import { ArticleCard } from "./articleCard";

export function Article({ user }) {
  const { listFilteredArticles } = useContext(NewsApiContext);
  const { title } = useParams();
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ title: title }),
    []
  );

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  if (!user.microsoft && !user.google) {
    return <Navigate to="/login" replace />;
  }

  return (
    <>
      {data.map((news) => (
        <>
          <ArticleCard key={news.title} news={news} />
        </>
      ))}
    </>
  );
}
