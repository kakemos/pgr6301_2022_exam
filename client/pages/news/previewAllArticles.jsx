import React, { useContext, useState } from "react";
import { NewsApiContext } from "../../newsApiContext";
import { useLoading } from "../../lib/useLoading";
import { FrontPageArticleCard } from "./frontPageArticleCard";

export function PreviewAllArticles() {
  const { listFilteredArticles } = useContext(NewsApiContext);
  const [category, setCategory] = useState("");
  const [categoryQuery, setCategoryQuery] = useState("");
  const { loading, error, data } = useLoading(
    async () => await listFilteredArticles({ category }),
    [category]
  );

  function handleSubmitQuery(e) {
    e.preventDefault();
    setCategory(categoryQuery);
  }

  if (loading) {
    return <div>Loading...</div>;
  }
  if (error) {
    return (
      <div>
        <h1>Error</h1>
        <div id="error-text">{error.toString()}</div>
      </div>
    );
  }

  return (
    <>
      <div>
        <form onSubmit={handleSubmitQuery}>
          <input
            type={"radio"}
            name={"filter-news"}
            value={"all"}
            onChange={() => setCategoryQuery("")}
          />
          <label htmlFor={"all"}>All</label>
          <input
            type={"radio"}
            name={"filter-news"}
            value={"general"}
            onChange={() => setCategoryQuery("general")}
          />
          <label htmlFor={"general"}>General</label>
          <input
            type={"radio"}
            name={"filter-news"}
            value={"fashion"}
            onChange={() => setCategoryQuery("fashion")}
          />
          <label htmlFor={"fashion"}>Fashion</label>
          <input
            type={"radio"}
            name={"filter-news"}
            value={"art"}
            onChange={() => setCategoryQuery("art")}
          />
          <label htmlFor={"art"}>Art</label>
          <input
            type={"radio"}
            name={"filter-news"}
            value={"travel"}
            onChange={() => setCategoryQuery("travel")}
          />
          <label htmlFor={"travel"}>Travel</label>
          <input
            type={"radio"}
            name={"filter-news"}
            id={"category-query"}
            value={"celebrity"}
            onChange={() => setCategoryQuery("celebrity")}
          />
          <label htmlFor={"celebrity"}>Celebrity</label>

          <button className={"filter-frontpage-article-button"}>Filter</button>
        </form>
      </div>

      {data.map((news) => (
        <FrontPageArticleCard key={news.title} news={news} />
      ))}
    </>
  );
}
