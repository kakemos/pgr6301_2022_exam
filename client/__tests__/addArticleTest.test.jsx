import ReactDOM from "react-dom";
import React from "react";
import { MemoryRouter } from "react-router-dom";
import { AddNewArticle } from "../pages/news/addNewArticle";
import { NewsApiContext } from "../newsApiContext";
import { act, Simulate } from "react-dom/test-utils";

describe("add movie component", () => {
  it("shows movies form", () => {
    const element = document.createElement("div");
    const user = { microsoft: { name: "Testy Testsen" } };
    ReactDOM.render(
      <MemoryRouter>
        <AddNewArticle user={user} />
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
    expect(
      Array.from(element.querySelectorAll("form label strong")).map(
        (e) => e.innerHTML
      )
    ).toEqual(["Title", "Intro", "Category", "Article"]);
  });

  it("should show article", () => {});

  it("refuse to add article with empty fields", () => {
    const addNewArticle = jest.fn();
    const location = new URL("https://www.example.com");
    delete window.location;
    window.location = new URL(location);
    const user = { microsoft: { name: "Testy Testsen" } };
    const element = document.createElement("div");
    ReactDOM.render(
      <NewsApiContext.Provider value={{ addNewArticle }}>
        <MemoryRouter>
          <AddNewArticle user={user} />
        </MemoryRouter>
      </NewsApiContext.Provider>,
      element
    );
    Simulate.change(element.querySelector(".form-input input"), {
      target: { value: "" },
    });
    Simulate.submit(element.querySelector("form"));
    expect(window.location.pathname).toContain("/");
  });

  // it("adds movie on submit", () => {
  //   const addNewArticle = jest.fn();
  //   const user = { microsoft: { name: "Testy Testsen" } };
  //   const title = "Test Article";
  //   const intro = "Test Intro";
  //   const article = "Test Article";
  //   const category = "Test Category";
  //   const author = user.microsoft.name;
  //   const element = document.createElement("div");
  //   ReactDOM.render(
  //     <NewsApiContext.Provider value={{ addNewArticle }}>
  //       <MemoryRouter>
  //         <AddNewArticle user={user} />
  //       </MemoryRouter>
  //     </NewsApiContext.Provider>,
  //     element
  //   );
  //   act(() => {
  //     Simulate.change(element.querySelector(".form-input input"), {
  //       target: { value: title },
  //     });
  //   });
  //   act(() => {
  //     Simulate.change(
  //       element.querySelector(".form-input:nth-of-type(2) input"),
  //       {
  //         target: { value: intro },
  //       }
  //     );
  //   });
  //   act(() => {
  //     Simulate.change(
  //       element.querySelector(".form-input:nth-of-type(4) input"),
  //       {
  //         target: { value: article },
  //       }
  //     );
  //   });
  //   Simulate.submit(element.querySelector("form"));
  //   expect(addNewArticle).toBeCalledWith({
  //     title,
  //     intro,
  //     category,
  //     article,
  //     author,
  //   });
  // });
});
