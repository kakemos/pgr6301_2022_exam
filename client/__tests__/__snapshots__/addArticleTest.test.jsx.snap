// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add article component shows new article form 1`] = `"<div>Loading...</div>"`;

exports[`add movie component shows movies form 1`] = `"<div>Loading...</div>"`;
